rust/src/ike/ike.rs

```diff
@@ -0,0 +1,628 @@
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                SCLogDebug!("Insufficient data while parsing IKE");
+                return AppLayerResult::err();
+            }
+            Err(_) => {
+                SCLogDebug!("Error while parsing IKE packet");
+                return AppLayerResult::err();
+            }
+        }
+    }
+
+    fn tx_iterator(
+        &mut self,
+        min_tx_id: u64,
+        state: &mut u64,
+    ) -> Option<(&IKETransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+}
+
+/// Probe to see if this input looks like a request or response.
+fn probe(input: &[u8], rdir: *mut u8) -> bool {
+    match parse_isakmp_header(input) {
+        Ok((_, isakmp_header)) => {
+            if isakmp_header.maj_ver == 1 {
+                if isakmp_header.resp_spi == 0 {
+                    unsafe {*rdir = STREAM_TOSERVER;}
+                }
+                return true
+            } else if isakmp_header.maj_ver == 2 {
+                if isakmp_header.min_ver != 0 {
+                    SCLogDebug!("ipsec_probe: could be ipsec, but with unsupported/invalid version {}.{}",
+                            isakmp_header.maj_ver, isakmp_header.min_ver);
+                    return false
+                }
+                if isakmp_header.exch_type < 34 || isakmp_header.exch_type > 37 {
+                    SCLogDebug!("ipsec_probe: could be ipsec, but with unsupported/invalid exchange type {}",
+                           isakmp_header.exch_type);
+                    return false
+                }
+                if isakmp_header.length as usize != input.len() {
+                    SCLogDebug!("ipsec_probe: could be ipsec, but length does not match");
+                    return false
+                }
+
+                if isakmp_header.resp_spi == 0 {
+                    unsafe {*rdir = STREAM_TOSERVER;}
+                }
+                return true
+            }
+
+            return false
+        },
+        Err(_) => {
+            return false
+        },
+    }
+}
+
+// C exports.
+export_tx_get_detect_state!(
+    rs_ike_tx_get_detect_state,
+    IKETransaction
+);
+export_tx_set_detect_state!(
+    rs_ike_tx_set_detect_state,
+    IKETransaction
+);
+
+/// C entry point for a probing parser.
+#[no_mangle]
+pub extern "C" fn rs_ike_probing_parser(
+    _flow: *const Flow,
+    _direction: u8,
+    input: *const u8,
+    input_len: u32,
+    rdir: *mut u8
+) -> AppProto {
+    if input_len < 28 {
+        // at least the ISAKMP_HEADER must be there
+        return unsafe { ALPROTO_FAILED };
```

Comment:
Could you add a comment that this is not `ALPROTO_UNKNOWN` because we are over UDP ?

Reply:
this is more coincidence, haven't debugged what happens after returning these constants :)